m2e-maven-runtime/org.eclipse.m2e.archetype.common/pom.xml

```diff
@@ -97,8 +97,8 @@
 						<![CDATA[
 							-exportcontents: \
 								META-INF.plexus;-noimport:=true;x-internal:=true,\
-								org.apache.maven.archetype.*;provider=m2e;mandatory:=provider;x-internal:=true,\
-								org.codehaus.plexus.velocity;provider=m2e;mandatory:=provider;x-internal:=true
+								org.apache.maven.archetype.*,\
```

Comment:
I'm not familiar with all that. Can you please suggest a PR that fixes it? Also not that the version should probably be the version of the Maven packages, not related to the version of the bundle.

Reply:
> Also not that the version should probably be the version of the Maven packages, not related to the version of the bundle

Exactly! That's why I think we should not change this just as is, as we might change the semantics here in an unpleasant way. The "provider" attribute for example is used to make sure we match exactly this bundle here and as it e.g. requires another bundle this might be important. @HannesWell has recently worked on that area any maybe he can better tell why we have bundled maven in that way (and not for example using individual 'official' bundles, or simply include everything in m2ecore at the first place...

Reply:
The generation of `Uses`-clauses are suppressed in the configuration of the bnd-maven-plugin in the parent pom using `-nouses: true`. As far as I can tell bnd only adds versions to exported-packages automatically if the package-version is specified in a package.info or similar. But we should be able to add the version 'manually'.

Some insights why providers are specified for the package-export/import can be found here https://bugs.eclipse.org/bugs/show_bug.cgi?id=403243 respectively 7e198c6ae5cffbd6fc45e0cb3b54492123d7e2e4

> The "provider" attribute for example is used to make sure we match exactly this bundle here and as it e.g. requires another bundle this might be important. @HannesWell has recently worked on that area any maybe he can better tell why we have bundled maven in that way (and not for example using individual 'official' bundles, or simply include everything in m2ecore at the first place...

These decisions were made way before my time so I can only guess here: I think one point was that the maven-targets were not available back then and embedding jars was the only way to make these maven-jars available as bundles, unless Orbit was used. But since this bundle not only contains the archtype-jars but also some dependencies maybe it was intended to have that exact that set of dependencies available and OSGi maybe would be too dynamic. Furthermore most of the dependencies don't have OSGi compliant Manifests so it would be some work to set them up.

For the archetype bundles I already considered to use the m2e-maven target and create a feature for them.
Having https://github.com/eclipse-m2e/m2e-core/issues/286 would help for that (I plan to implement that but did not have the time yet). Besides that it likely will result in many generated Manifests.

I think we should not do that for the actual maven-runtime jars because a maven version is usually a fixed set of jars and OSGi could be too dynamic here. But the archetype is only used in the UI so I think that's ok.

Btw. There are some more versions available to which we could update:
https://search.maven.org/artifact/org.apache.maven.archetype/archetype-common
But lets do that after 2.0.0 to have a bit more time to find regressions.

Reply:
```suggestion
							-nouses: false
							-exportcontents: \
								META-INF.plexus;-noimport:=true;x-internal:=true,\
								org.apache.maven.archetype.*;version="${archetype-common.version}"
```

Would generate uses and versions.
Besides that this change would fix some access resection warnings. I first considered to add `;x-friends:="org.eclipse.m2e.core,org.eclipse.m2e.core.ui"`. But making it public would fix that of course too.

Reply:
> > -nouses: false
> 
> Export-Package are not generated with this option.

I cannot confirm this.
But the suggested replacement has to be applied for the whole `exportcontents` and not only for the line. I just was not able to change the range for the suggestion in this thread.

Reply:
> > I cannot confirm this.
> 
> You can run
> 
> ```
> mvn generate-sources -f m2e-maven-runtime -Pgenerate-osgi-metadata -Dtycho.mode=maven
> ```

I know :)
I meant that when I run that command respectively the corresponding launch configuration contained in this repo it works fine if the change is applied correctly.

Reply:
You are right. I have set
```
-nouses: false \
```
instead of
```
-nouses: false
```